### Task 1
Data Cleaning Challenge on Sales Data.


In [7]:
import pandas as pd
sales_2017=pd.read_csv("Sales-Transactions-2017.csv.csv")
sales_2018=pd.read_csv("Sales-Transactions-2018.csv.csv")
sales_2019=pd.read_csv("Sales-Transactions-2019.csv.csv")
sales_edited=pd.read_csv("Sales-Transactions-Edited.csv.csv")

### Concatinating the data

In [160]:
sales_fulldata=pd.concat(objs=[sales_2017,sales_2018,sales_2019])

### Checking the shapes of two tables

In [10]:
sales_edited.shape

(95562, 6)

In [17]:
sales_fulldata.shape

(111206, 9)

Need to clean the rows to 95562 and columns to 6



### Checking the null totals of the two table

In [13]:
sales_edited.isnull().sum()  #except rate there are no nulls in any column

Date       0
Voucher    0
Party      0
Product    0
Qty        0
Rate       1
dtype: int64

In [19]:
sales_fulldata.isnull().sum()

Date               12591
Voucher            12557
Party                 40
Product            12591
Qty                12557
Rate               12558
Gross              12558
Disc              105609
Voucher Amount     83646
dtype: int64

In [ ]:
Need to remove the nulls after checking how many duplicates and why null is populated
What i can observe is null values are same for (Date, Product 12591 >111206 rows ) , (Voucher,Quantity 12557 ?> 111206 rows) others are different
Out of 111206 rows in disc 105609 are null
Out of 111206 rows in Voucher Amount 83646 are null
40 party are null

### Checking data types and descriptive statistics of the two table

In [21]:
sales_edited.dtypes

Date        object
Voucher      int64
Party       object
Product     object
Qty          int64
Rate       float64
dtype: object

In [22]:
sales_fulldata.dtypes

Date              object
Voucher           object
Party             object
Product           object
Qty               object
Rate              object
Gross             object
Disc              object
Voucher Amount    object
dtype: object

Gross, Disc and Voucher amount is not there in original data
Date should be in date format
Voucher in int64
Qty in int64
Rate in float64

### Observing first few rows of both the table?

In [29]:
sales_edited.head(20)

,Date,Voucher,Party,Product,Qty,Rate
0,1/4/2017,1,SOLANKI PLASTICS,DONA-VAI-9100,2,1690.00
1,1/4/2017,1,SOLANKI PLASTICS,LITE FOAM(1200),6,1620.00
2,1/4/2017,2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23.00
3,1/4/2017,2,SARNESWARA TRADERS,LITE FOAM(1200),6,1620.00
4,1/4/2017,2,SARNESWARA TRADERS,DONA-VAI-9100,5,1690.00
5,1/4/2017,2,SARNESWARA TRADERS,CLASSIC ENJOY(750),1,1320.00
6,1/4/2017,898,LOCK,VISHNU 250ML,100,30.00
7,1/4/2017,898,LOCK,BLACK DOG-350ML,100,26.00
8,1/4/2017,2497,VAMSI KRISHNA FANCY,LOOSE ITEMS,1,800.00
9,1/4/2017,9263,VAMSI KRISHNA FANCY,LOOSE ITEMS,1,280.00


In [30]:
sales_fulldata.head(20)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
5,1/4/2017,Sal:2,SARNESWARA TRADERS,CLASSIC ENJOY(750),1,"1,320.00","1,320.00",NaN,NaN
6,1/4/2017,Sal:898,Lock,Vishnu 250ml,100,30,"3,000.00",100,"5,400.00"
7,1/4/2017,Sal:898,Lock,BLACK DOG-350ML,100,26,"2,600.00",100,NaN
8,,,khader vali late en,,,,,,
9,,,try,,,,,,


In voucher sal: is removed and only the numbers are kept
Many dummy entrey are there or unwanted rows are there
Not sure why Gross, disc and voucher amount were removed need to study those three columns first
Discrete features are date, party, and product others seems to be continous features. Grouping can be done with these discrete deatures

### Observing duplicate values

In [95]:
sales_fulldata[sales_fulldata.duplicated()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
14,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
17,,,dummy entry,,,,,,
39,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
41,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
43,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
19143,NaN,NaN,ET,NaN,NaN,NaN,NaN,NaN,NaN
19162,,,vinayaka sweets-kot,,,,,,
19163,,,hagunta,,,,,,
19172,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Total 14582 rows are duplicates out of 111206, if removed it will become 96624 in cleaned sales_edited data there are only 95562

1. Need to clean the rows to 95562 and columns to 6 as per sales_edited
2. Need to remove the nulls after checking how many duplicates and why null is populated
3. Null values are same for (Date, Product 12591) , (Voucher,Quantity 12557) others are different
4. Out of 111206 rows in disc 105609 are null
5. Out of 111206 rows in Voucher Amount 83646 are null
6. 40 party rows are null
7. Gross, Disc and Voucher amount is not there in original sales_edited data
8. Date should be in date format
9. Voucher in int64 format
10. Qty in int64 format
11. Rate in float64 format
12. In voucher sal: is removed and only the numbers are kept in sales_edited
13. Many dummy entrey are there or unwanted rows are there
14. Not sure why Gross, disc and voucher amount were removed need to study those three columns first
15. Discrete features are date, party, and product others seems to be continous features. Grouping can be done with these discrete deatures
16. Total 14582 rows are duplicates out of 111206, if removed it will become 96624 in cleaned sales_edited data there are only 95562
17. 



### Removing Duplicates from the sales_fulldata?

In [161]:
sales_fulldata.drop_duplicates(inplace=True)
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN
19171,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [40]:
sales_fulldata[sales_fulldata.duplicated()] #no duplicates in the new data

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount


### Find total null, count and unique values in the dataset?

In [74]:
sales_fulldata["Party"].isnull().sum()

np.int64(35)

In [52]:
sales_fulldata.isnull().sum()

Date                204
Voucher             170
Party                35
Product             204
Qty                 170
Rate                171
Gross               171
Disc              92951
Voucher Amount    70984
dtype: int64

In [55]:
sales_fulldata.count()

Date              96420
Voucher           96454
Party             96589
Product           96420
Qty               96454
Rate              96453
Gross             96453
Disc               3673
Voucher Amount    25640
dtype: int64

In [54]:
sales_fulldata.nunique()

Date                836
Voucher           10044
Party              1994
Product             867
Qty                 546
Rate               1109
Gross              2548
Disc                371
Voucher Amount     7149
dtype: int64

### Finding if "Party","Product","Date" all three data are null, if then it can be removed?

In [101]:
sales_fulldata.isna().sum()

Date                204
Voucher             170
Party                35
Product             204
Qty                 170
Rate                171
Gross               171
Disc              92951
Voucher Amount    70984
dtype: int64

### 35 null values in party have other categories data and product also null so this 35 columns will be dropped first

In [167]:
sales_fulldata[sales_fulldata["Party"].isna()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount


In [168]:
sales_fulldata.dropna(subset="Party",inplace=True)
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


### Now i am checking null values again in all the columns- which shows 0 in party and 169 to 170 in other columns

In [169]:
sales_fulldata.isna().sum()

Date                169
Voucher             169
Party                 0
Product             169
Qty                 169
Rate                170
Gross               170
Disc              92950
Voucher Amount    70983
dtype: int64

### When i again see the dispaly of this 169 null columns it seems like we cannot use these rows as we cannot impute any data. Deleting this columns as well

In [170]:
sales_fulldata[sales_fulldata["Date"].isna()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
11,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
53,NaN,NaN,KR,NaN,NaN,NaN,NaN,NaN,NaN
77,NaN,NaN,IC CENTRE,NaN,NaN,NaN,NaN,NaN,NaN
79,NaN,NaN,LAKSHMIMEDICALS,NaN,NaN,NaN,NaN,NaN,NaN
104,NaN,NaN,LS VGIRI,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
16853,NaN,NaN,VA-VGIRI,NaN,NaN,NaN,NaN,NaN,NaN
16890,NaN,NaN,ADY MIX,NaN,NaN,NaN,NaN,NaN,NaN
17015,NaN,NaN,HARI-SRIKALAHASTHI,NaN,NaN,NaN,NaN,NaN,NaN
18033,NaN,NaN,RU,NaN,NaN,NaN,NaN,NaN,NaN


In [126]:
sales_fulldata.dropna(subset="Date",inplace=True)

In [171]:
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


In [128]:
sales_fulldata.isna().sum()

Date                  0
Voucher               0
Party                 0
Product               0
Qty                   0
Rate                  1
Gross                 1
Disc              92781
Voucher Amount    70814
dtype: int64

In [172]:
sales_fulldata.head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
5,1/4/2017,Sal:2,SARNESWARA TRADERS,CLASSIC ENJOY(750),1,"1,320.00","1,320.00",NaN,NaN
6,1/4/2017,Sal:898,Lock,Vishnu 250ml,100,30,"3,000.00",100,"5,400.00"
7,1/4/2017,Sal:898,Lock,BLACK DOG-350ML,100,26,"2,600.00",100,NaN
8,,,khader vali late en,,,,,,
9,,,try,,,,,,


In [180]:
sales_fulldata[(sales_fulldata[["Date","Voucher","Qty"]]==" ")].count()

Date              1133
Voucher           1133
Party                0
Product              0
Qty               1131
Rate                 0
Gross                0
Disc                 0
Voucher Amount       0
dtype: int64

### Many blanks are there so changing those blanks to nan using numpy and replace

In [ ]:
import numpy as np
sales_fulldata.replace(r'^\s*$', np.nan, regex=True, inplace=True) 

### Found some more null values now need to remove these

In [137]:
sales_fulldata.isna().sum()

Date               1133
Voucher            1133
Party                 0
Product            1133
Qty                1131
Rate               1132
Gross              1134
Disc              93914
Voucher Amount    71947
dtype: int64

### Displaying these 1133 rows to see whether all have null values

In [166]:
sales_fulldata[sales_fulldata["Date"].isna()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
11,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
53,NaN,NaN,KR,NaN,NaN,NaN,NaN,NaN,NaN
77,NaN,NaN,IC CENTRE,NaN,NaN,NaN,NaN,NaN,NaN
79,NaN,NaN,LAKSHMIMEDICALS,NaN,NaN,NaN,NaN,NaN,NaN
104,NaN,NaN,LS VGIRI,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
16853,NaN,NaN,VA-VGIRI,NaN,NaN,NaN,NaN,NaN,NaN
16890,NaN,NaN,ADY MIX,NaN,NaN,NaN,NaN,NaN,NaN
17015,NaN,NaN,HARI-SRIKALAHASTHI,NaN,NaN,NaN,NaN,NaN,NaN
18033,NaN,NaN,RU,NaN,NaN,NaN,NaN,NaN,NaN


### Deleting these 1133 null values

In [146]:
sales_fulldata.dropna(subset="Date",inplace=True)
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


In [165]:
sales_edited.nunique()


Date         835
Voucher    10042
Party        712
Product      866
Qty          354
Rate        1073
dtype: int64

In [164]:
sales_fulldata.nunique()

Date                836
Voucher           10043
Party              1994
Product             867
Qty                 512
Rate               1075
Gross              2514
Disc                337
Voucher Amount     7115
dtype: int64

In [76]:
sales_fulldata.dropna(subset=["Party","Product","Date"],how="all").count()

Date              96420
Voucher           96420
Party             96589
Product           96420
Qty               96420
Rate              96419
Gross             96419
Disc               3639
Voucher Amount    25606
dtype: int64

In [66]:
sales_fulldata.dropna()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN
19171,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [85]:
sales_fulldata["Party"].dropna()

0          SOLANKI PLASTICS
1          SOLANKI PLASTICS
2        SARNESWARA TRADERS
3        SARNESWARA TRADERS
4        SARNESWARA TRADERS
                ...        
19167             K.SRIHARI
19168             K.SRIHARI
19169             K.SRIHARI
19170             K.SRIHARI
19171             K.SRIHARI
Name: Party, Length: 96589, dtype: object